substrate/primitives/api/proc-macro/src/impl_runtime_apis.rs

```diff
@@ -81,6 +82,19 @@ fn generate_impl_call(
 	let ptypes = params.iter().map(|v| &v.1);
 	let pborrow = params.iter().map(|v| &v.2);
 
+	// usage of `Self` is not allowed in the scope of the trait implementation
+	params.iter().try_for_each(|v| {
+		if let Type::Path(path) = &v.1 {
```

Comment:
But this only checks it for params, and not the function body?  
Or is it allowed there? Otherwise i would rather check it with [`syn::visit`](https://docs.rs/syn/latest/syn/visit/index.html).